migrations: Remove disallowed characters from topics.

mateuszmandera opened the pull request:
> Following b3c58f454f0dce8a88c696305945bac41f9786bc, we want to clean up
old topics that may contain the disallowed characters. The Message table
is large, so we go in batches, making sure we limit topic fetches and
UPDATE query to no more than BATCH_SIZE Message rows per query.


It seemed tempting to try to limit the number of topics processed per batch, rather than going by Message id ranges, but it complicates reasoning about the logic (huge UPDATE if the topic is super large for example, the fact that the ordering of topics can change as a result of removing some characters from a topic etc.)

mateuszmandera commented on zerver/migrations/0371_invalid_characters.py:
```diff
@@ -0,0 +1,54 @@
+
+def fix_topics(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Message = apps.get_model("zerver", "Message")
+    BATCH_SIZE = 10000
```
Does 10,000 seem right?

timabbott replied:
Yeah, that should be fine.

chrisbobbe commented:
Thanks! See my question about using "(no topic)" for topics that would otherwise be "", at https://chat.zulip.org/#narrow/stream/141-kandra-ops/topic/Special.20characters.20in.20topics/near/1294838.

timabbott commented on zerver/migrations/0371_invalid_characters.py:
```diff
@@ -0,0 +1,54 @@
+                    ]
+                )
+                if fixed_topic == topic:
+                    continue
```
I think @chrisbobbe is right that we should replace `""` with `"(no topic)"` here.  We need to be careful, though, to avoid doing that sort of rewrite for PMs; I guess since PMs don't have topics, we can do that by just doing the `if fixed_topic == ""` check after this continue.

timabbott commented on zerver/migrations/0371_invalid_characters.py:
```diff
@@ -0,0 +1,54 @@
+    ]
+
+    operations = [
+        migrations.RunPython(fix_topics, reverse_code=migrations.RunPython.noop),
```
Let's name the migration file `invalid_characters_in_topics`; it seems worth being a bit more specific.

mateuszmandera commented:
Renamed the migration and added the extra if:
```
                if fixed_topic == topic:
                    continue

                if fixed_topic == "":
                    fixed_topic = "(no topic)"
```

timabbott commented:
This is great; I added a bit more of our standard progress logging, test-deployed it on chat.zulip.org (fixing 63 messages, apparently) and merged this. Thanks @mateuszmandera!
